### ISSUE TITLE
Feature Request: MSI package for Enterprise deployment

### ISSUE THREAD

**Issue by jNizM** [notepad-plus-plus/notepad-plus-plus#2326]
Hey Don Ho,

is there a way to realease also a MSI package for Enterprise deployments?
So it can be deployed in Enterprise via Active Directory / Group Policy?

greets
jNizM

**Comment by yamlfd** [comment]
Full fledged MSI installer is here on my site https://www.hass.de/content/notepad-msi-package-enterprise-deployment-available. I shared the source code with Don 2.5 years ago. Still no movement.

**Comment by yamlfd** [comment]
Refactored all my MSI build scripts. v7.0.0 is also available now under https://www.hass.de/content/notepad-msi-package-enterprise-deployment-available

**Comment by willemdh** [comment]
@yamlfd  Thanks a lot! I've been using your msi's for some time now.

**Comment by Ulyssys** [comment]
Just wanted to voice support for the MSI install package. NSIS is okay, but an MSI is far preferable for any kind of mass Windows deployment and less error-prone when automated.

@yamlfd Would you be willing to share the source for your MSI build scripts publicly?

**Comment by yamlfd** [comment]
I shared several previous versions with the maintainer, but he has not taken over the job. I wish we would stop with nsis shit.

The msi's can be downloaded from www.hass.de

**Comment by Ulyssys** [comment]
@yamlfd Are you willing to share the _source code_ for the build scripts publicly? Spin up a new repo forking the existing code on Github, then the community can benefit.

Your completed MSIs are useful, but it would be proper to open the source so that it can be easily maintained in a repo, verified independently, continued by other authors, etc.

**Comment by jNizM** [comment]
Hey Don Ho.
Any updates here for an official msi-installer?
greets

**Comment by Ulyssys** [comment]
Pinging @alexhass and @donho hopefully for some followups

**Comment by alexhass** [comment]
Don does not answer. It‘s like a black hole. No idea why we cannot get it forward.

**Comment by Ulyssys** [comment]
Hi @alexhass,

To be fair, you did not answer the question about providing the source for your work as per my previous comment, either. It can be easy to overlook sometimes!

**Comment by alexhass** [comment]
I will share it when it will be added to the project, but it looks like Don like dead horses like nsis more and do not understand that a msi wrapper is wrong direction. It is more the other way around... build an exe with an embedded msi if an exe is required.

I do not know why you need to work on my msi code. Your help is currently not needed from my point of view. If you have any special needs you can share your ideas and I may add this on payed basis.

**Comment by Ulyssys** [comment]
@alexhass 

It's not about my personal need to work on your code, but rather that the community can benefit from it, and provide benefit to it.

I am sure you understand how open source works, the nature of the GPL license that Notepad++ uses, as well as the current instructions on submitting new code, [linked here from the homepage](https://notepad-plus-plus.org/contribute/).

So, I am curious why you think the code shouldn't be public if you plan on contributing it to the project anyway?

It seems strange that you so strongly acknowledge the community's need for an improved solution, yet denigrate Don when a fork/pull request is a simple enough task and would get it adopted much more quickly by the community. Are you only interested in selling your code?

**Comment by alexhass** [comment]
I shared my code first time long before the github repository existed. No need to deningate Don, he made it on it‘s own. Only review the queue here or send him an email. No answers, no solutions, no progress, no way to commit code to move the project forward.

My main main reason is frustration. After nearly 3 years of frustration on this project I‘m not very motivated - without changes to how issues are handled. I have bugs opened 2 years ago where stupid notepad request UAC on a plugin install to appdata onna network drive. Zero feedback and none of my users can install plugins since this day. I cannot solve myself as I do not develop notepad. Open source or not, this issue queue here is waste of time. Nobody has a focus on resolving the issues. It just grow and grow.

I do not sell the code. When I started I hoped to help the project and get the work away from my personal desk. I still have to build the setup as I need to deploy it. It looks like Don has no idea about software deployment. Otherwise he would rip nsis. So - change something, fix the serious bugs and move the project forward and people may become motivated again.

**Comment by willemdh** [comment]
Or just use https://chocolatey.org/packages/notepadplusplus ?

**Comment by alexhass** [comment]
This is not an msi.

**Comment by JeromeFa** [comment]
Hi Alex,
Will you still maintain your msi notepad++ solution or will you stop ?
Thanks
Jerome

**Comment by alexhass** [comment]
Yes I do.

Gruss
Alex

**Comment by JeromeFa** [comment]
Great thanks

**Comment by jNizM** [comment]
Hey DonHo,
is there any news on that? I (and many others) would be happy about this feature.

greets

**Comment by rrsit** [comment]
+1 for native msi.

**Comment by Conan1989** [comment]
+1 also :)

**Comment by tuomaskuittinen** [comment]
+1 for .msi also

**Comment by RobinFlikkema** [comment]
+1

**Comment by SecT0uch** [comment]
+1

**Comment by SecT0uch** [comment]
++1

**Comment by SecT0uch** [comment]
7.7 released, still no msi :(

**Comment by bluikko** [comment]
@SecT0uch indeed, very sad. I am thinking about starting to use PDQ Deploy due to so many OSS projects not having MSI installers - even VSCode has not got one.

MSI installer would be really useful.

Additionally it looks like @alexhass has given up, 7.6.3 (4 months old) is the latest at www.hass.de

**Comment by alexhass** [comment]
Nope. I‘m only limited in time and building installers takes time. 

I have build vscode mai setup, but the project maintainers refused it later. They do not understand customer and enterprise needs.

**Comment by DomBros** [comment]
+1

**Comment by bluikko** [comment]
@alexhass Thank you for making MSI for 7.7.1! Your work is very much appreciated. I hope you keep at it.

Is there any possibility to get it digitally signed?

**Comment by alexhass** [comment]
I wish I coud, but I do not have the certificate from Don Ho and do not have a free account where I can sign. Paying for signing is therefore only possible with sponsorship.

**Comment by janparttimaa** [comment]
@alexhass I have tried your MSI packeted Notepad++, it's really great and handy!

I also agree for this. Official MSI-packet would be great for Enterprise Deployments. @donho Is there any reasons why there has been no official MSI-packet? Many program publishers has been released both installers; MSI and exe, just like 7zip have been done:

https://www.7-zip.org/download.html

I think if making co-operation with @alexhass, official MSI-installer would not be hard to create and it would be released quickly.

**Comment by bluikko** [comment]
Agreed. The contribution by @alexhass is seriously good. I hope he keeps updating it sometimes - I'm happy as long as new MSIs are created for future security updates.

**Comment by janparttimaa** [comment]
@alexhass Do you wanna make pull request of MSI-installer?

**Comment by mimi89999** [comment]
@janparttimaa Firefox also released a MSI installer recently.

**Comment by c3rberus** [comment]
Can we get MSI package for Notepad++? This is seriously needed for ease of automation.

**Comment by NemoDima** [comment]
MSI provides the ability to use Powershell DSC to automate environment preparation.
I'll be appreciate if Author provide it.

**Comment by tb-mtg** [comment]
I noticed this is tagged as accepted, is there any progress on this?

**Comment by bluikko** [comment]
I don't see why there would need to be 2 installers both EXE and MSI. If I'm not mistaken MSI can do anything the EXE installer does, plus is a proper managed install that can be used with deployment tools.

I've been very happy with Hass' packages but it would be great to have an official MSI that's always up to date.

This issue looks to be one of the most popular open issues in GitHub which reflects the high importance of the request!

**Comment by c3rberus** [comment]
@alexhass will you be releasing an MSI for 8.x soon? :)

**Comment by bluikko** [comment]
> @alexhass will you be releasing an MSI for 8.x soon? :)

I would rather wait for .1 or .2 release... or better yet **an official MSI** but seems that is too much to ask.

**Comment by shirou93** [comment]
This project is deprecated? https://www.hass.de/content/notepad-msi-package-enterprise-deployment-available

**Comment by bluikko** [comment]
@shirou93 In the past he has not created packages for every version - just sometimes if there is something important. I've not followed if there are any security vulnerabilities found in 7.9.5, if not then I am not surprised there has not been a new version.

In the long-term it should be the npp project that builds MSIs, too bad it still has not happened!

**Comment by alexhass** [comment]
No, just other priorities.

**Comment by matkoniecz** [comment]
>  They do not understand customer and enterprise needs.

Note that it is hard to say that N++ has customers given that it is available for free.

Note that author of this project appears to be really good at processing submitted pull requests - and none of six currently open is implementing this ( https://github.com/notepad-plus-plus/notepad-plus-plus/pulls )

https://github.com/notepad-plus-plus/notepad-plus-plus/pulls?q=is%3Apr+MSI - single was opened two years ago. I have not reviewed why it was closed without merging.

It seems to me that noone is interested enough to actually implement it and open a pull request.

**Comment by jonesmz** [comment]
Hi, I was trying to deploy via an active directory group policy object today, but couldn't because Notepad++ isn't distributed with an MSI installer.

Please consider including this in an upcoming release.

**Comment by bluikko** [comment]
> I would very much like to see this implemented. Other OSS software is able to accomplish this, so I'm sure there is an effective/efficient way to do it, but this isn't within my areas of direct knowledge.

The unofficial MSI from Hass, listed in a comment above, has worked well for me, It does not get updated regularly but it's been fine for my uses.

**Comment by alexhass** [comment]
@donho there are double slashes in one folder. Never use regsrv to register the shell dll. Do not use custom actions whenever possible. There are native ways… you got my msi package years ago… so you can copy from there. I do not see how enterprise can disable automatic updates to integrate into enterprise deployment like intune.

**Comment by donho** [comment]
@alexhass 
> there are double slashes in one folder

Well catched. It'll be fixed.

> Never use regsrv to register the shell dll. Do not use custom actions whenever possible. There are native ways… 

If you're talking about heat.exe, I've tried and it generated some "harvest problem" warning.
The custom actions & regsvr works fine - so I see no reason to not use it.

> There are native ways… you got my msi package years ago… so you can copy from there.

I cannot find it anymore. Could you provide the link so I can download it?

> I do not see how enterprise can disable automatic updates to integrate into enterprise deployment like intune.

Indeed. I'll wait & see the feedback from the IT teams for the future enhancement.

**Comment by alexhass** [comment]
I’m not talking about heat. Can share a code snippet later. Again, never use custom actions if there is no need and a native way. This is bad practice. Running executables via custom actions is absolutly bad practice.

I’m missing that installation components are features like it is with the normal installer. This allows selection of every single feature.

Without beeing able to disable updates the setup cannot used. This is no enhancement, it is crucial.

**Comment by bluikko** [comment]
> Without beeing able to disable updates the setup cannot used. This is no enhancement, it is crucial.

Agreed. The raison d'etre for MSI is managed configurations for enterprises. Disabling updates is a requirement for that, simple as.

**Comment by janparttimaa** [comment]
Why not just use exe installer with silent switch and switch to disable updater? [https://npp-user-manual.org/docs/command-prompt/#installer-options](https://npp-user-manual.org/docs/command-prompt/#installer-options)

**Comment by alexhass** [comment]
> Why not just use exe installer with silent switch and switch to disable updater? https://npp-user-manual.org/docs/command-prompt/#installer-options

Exe setup are big shit and cannot customized.

MSI is the ONLY supported by AD deployment and works always a lot better than NSIS or INNO setups in all other situations.

**Comment by xomx** [comment]
@alexhass 

I've read a few of your posts here and I have the following comment - instead of childish shouts like "Exe setup are big shit", it would be much productive to contribute here using a standard Pull-Request, as is common in the open-source community (if you really need such a new feature and are capable to implement it).

Otherwise it's from your side only a wasting of your time & energy. Now, when the N++ author undertaken (for him probably unnecessary) creating of the msi N++ installer variant, it's much easier for ANYONE to simply use the existing N++ base msi support and extend it to all the other sysadmin msi-stuff needs.

No offense, but remember that this is an opensource SW, no one is a paying customer here. So while it's good to want things, it's always better to contribute if someone wants it right away or precisely according to his/her ideas.

**Comment by alexhass** [comment]
@xomx Just for you.

I’m the guy who created and maintains the MSI setup for notepad++ for about ~10 years. Donho got the raw code in past from me, but he never integrated it into the project.

Out of nothing / without collaboration for years - he pushed some unready and highly unusable code to the repro… rethink what you said now, please.

My MSI replaces his NSIS setup 1:1 in perfect manner. It has TONS of other features and he never said a word why we stick at NSIS crap.

**Comment by xomx** [comment]
@alexhass 

> I’m the guy who created and maintains the MSI setup for notepad++ for about ~10 years.

I know.

> Donho got the raw code in past from me, but he never integrated it into the project.

IDK that. I can't speak for him, but did you consider that maybe he just didn't have the time or the necessary knowledge or just didn't consider it worthwhile from his own POV/interests?

> Out of nothing / without collaboration for years - he pushed some unready and highly unusable code to the repro… rethink what you said now, please.

I don't have to re-think. You should have created here your own parallel (to the existing "NSIS crap" one) msi-installer pull-request (PR) and then discussed any implementation details/issues within it. And you didn't even have to look at the existing NSIS exe-installer, you could have simply created a parallel installer variant and if it was better and more universal, then over time it would probably completely replace the original NSIS-one.

People often thinks that when a feature request (FR) here hasn't been fulfilled yet, it's automatically considered as rejected or something. They don't consider at all any other possibilities. But in opensource it usually works this way: either you contribute indirectly with ideas, bug reports etc. or directly via the PRs. With the indirect support, you are then dependent on the free will/time of the other devs. With the PR way, you have more or less a direct control about the FR implementation and what's more - you can have the requested feature right away whenever you want to.

> My MSI replaces his NSIS setup 1:1 in perfect manner. It has TONS of other features and he never said a word why we stick at NSIS crap.

You are evidently more than capable in the msi-area, so I repeat - why didn't YOU create the parallel MSI-installer PR `HERE` from the start? You don't need any explicit permission for creating a PR here. If it's good, then there is a high probability that the project maintainer will accept it.

It would be of course another matter if for some reason you don't want or cannot participate in such an opensource development. If that's the case, then please ignore my comments. Otherwise, why wait next "10 years" for someone else to do it right here when you can do it (and probably better, as you say)?

**Comment by donho** [comment]
@alexhass 
I think I owe you an explanation.

The reason the MSI solution was previously rejected is based on the following arguments (that I've provided):
1. the difficulty of surpporting multiple languages in MSI.
2. MSI package built with WiX has had a visual glichy during the installation (a common issue across all apps using WiX-based MSI). This might be resolved in WiX v6 - to be verified.

These shortcomings prevent MSI versions from replacing NSIS installer in the past. Maintaining 2 different types of installers would be tedious, not to mention the confusion during uninstallation if a user installs Notepad++ with NSIS then switch to MSI.

As Notepad++ project matures, I've had more time to reconsider MSI deployment.
The feature I have in mind is clear: a basic and simple MSI, used solely for IT deployment - nothing more.

Most importantly, it's an MSI I can maintain independently, without relying on anyone else. That's why I've started building it from scratch myself.

The stub now exits and is part of project. You are welcome to contribute (via PR), provided you follow the controbution guidelines:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md

**Comment by alexhass** [comment]
> the difficulty of surpporting multiple languages in MSI.

What is and was incorrect. I guess you never looked into the code. I have implemented ***all*** languages with auto-detection. You can download the MSI from my site and test it. It works 100%. The language files are just nice XML files, not INI files.

> MSI package built with WiX has had a visual glichy during the installation (a common issue across all apps using WiX-based MSI). This might be resolved in WiX v6 - to be verified.

I do not understand this comment. Can you explain, what you are talking about, please? My setup works so much better than the NSIS setups and there are no visual issues... and more features than your NSIS setup. Please share an image of what you are talking, please.

> not to mention the confusion during uninstallation if a user installs Notepad++ with NSIS then switch to MSI.

You may remember - I implemented an silent upgrade from NSI zu MSI. This is no issue. I fully agree, maintaining two setups is bad.

> The feature I have in mind is clear: a basic and simple MSI, used solely for IT deployment - nothing more.

Not the best idea. You should move 100% to MSI.

This is the way how to add the context menu DLL properly:

```
<DirectoryRef Id="CONTEXTMENU">
      <?if $(env.Platform)=x86?>
        <Component Id="NppShell32" Guid="*">
          <Condition>NOT VersionNT64</Condition>
          <File Id="NppShell.dll" KeyPath="yes" Source="$(var.ReproDir)\contextMenu\NppShell.dll">
            <Class Id="{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}" Context="InprocServer32" Description="!(loc.ProductName)" ThreadingModel="apartment" />
          </File>
          <RegistryValue Root="HKCR" Key="*\shellex\ContextMenuHandlers\Notepad++" Value="{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}" Type="string" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="Title" Value="Edit with &amp;Notepad++" Type="string" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="Path" Value="[APPLICATIONFOLDER]notepad++.exe" Type="string" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="Custom" Value="" Type="string" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="ShowIcon" Value="1" Type="integer" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="Dynamic" Value="1" Type="integer" Action="write" />
          <RegistryValue Root="HKCR" Key="CLSID\{00F3C2EC-A6EE-11DE-A03A-EF8F55D89593}\Settings" Name="Maxtext" Value="25" Type="integer" Action="write" />
        </Component>
      <?endif?>
      <!--
      Ignore the validation errors about this. The file should go
      into NPP folder below "ProgramFiles64Folder", but we do not care
      if NPP is not installed on x64 and we do not like to create a x64
      folder only for the shell DLL.
      -->
      <Component Id="NppShell64" Guid="*" Win64="yes">
        <Condition>VersionNT64</Condition>
        <File Id="NppShell64.dll" KeyPath="yes" Source="$(var.ReproDir)\contextMenu\NppShell64.dll">
          <Class Id="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Context="InprocServer32" Description="!(loc.ProductName)" ThreadingModel="apartment" />
        </File>
        <RegistryValue Root="HKCR" Key="*\shellex\ContextMenuHandlers\Notepad++64" Value="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Title" Value="Edit with &amp;Notepad++" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Path" Value="[APPLICATIONFOLDER]notepad++.exe" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Custom" Value="" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="ShowIcon" Value="1" Type="integer" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Dynamic" Value="1" Type="integer" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Maxtext" Value="25" Type="integer" Action="write" />
      </Component>
    </DirectoryRef>
```

**Comment by donho** [comment]
@alexhass 
Thank you for your code.
However, I managed to integrate it into current ws, compiled with WiX v6, and no compiling error: 

```xml
	<ComponentGroup Id="ContextMenuComponentGroup" Directory="ContextMenuFolder">
	  <Component Id="cmp_NppShellDll" Guid="{D6083500-DFD2-445D-B4FF-49EDF80A6A84}" Bitness="always64">
		<File Source="..\..\bin64\NppShell.x64.dll" Name="NppShell.dll" KeyPath="yes" >
          <Class Id="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Context="InprocServer32" Description="Notepad++ Context Menu" ThreadingModel="apartment" />
        </File>
		<RegistryValue Root="HKCR" Key="*\shellex\ContextMenuHandlers\Notepad++64" Value="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Title" Value="Edit with &amp;Notepad++" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Path" Value="[APPLICATIONFOLDER]notepad++.exe" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Custom" Value="" Type="string" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="ShowIcon" Value="1" Type="integer" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Dynamic" Value="1" Type="integer" Action="write" />
        <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Maxtext" Value="25" Type="integer" Action="write" />
	  </Component>
	  
	  <Component Id="cmp_NppShellMsix" Guid="{EB6E5606-60D1-4C69-8C9B-C0B12A19C6CE}">
		<File Source="..\..\bin64\NppShell.msix" KeyPath="yes" />
	  </Component>
	</ComponentGroup>

```

However, under my Windows11, after the installation, there's no Notepad++ context menu entry at all.
If you see any error in the above code cause the absent of context menu entry, feel free to let me know.

**Comment by gjibrane** [comment]
Hello,
is there a way to use installer options with msi ?
I mainly need noUpdater.

**Comment by donho** [comment]
@gjibrane
> I mainly need noUpdater.

I have just done this option! `msiexec /i Notepad++.msi NOUPDATER=1`
Please check here:
https://download.notepad-plus-plus.org/MISC/npp.Installer.x64.msi

**Comment by gjibrane** [comment]
@donho 
Thank you.
I just used your msi to update a nsis 8.5.4 with sccm.
It worked but I'm not sure NOUPDATER worked because update option is still available in '?'.

I had an instant reboot, no warning at all.
Is there a NOREBOOT option ?

**Comment by donho** [comment]
@gjibrane 
> It worked but I'm not sure NOUPDATER worked

I have tested `msiexec /i Notepad++.msi NOUPDATER=1`and it works for me.

> because update option is still available in '?'.

I don't follow you. Could you elaborate it?

> Is there a NOREBOOT option ?

There's no such option, but normal, Notepad++ MSI doesn't reboot, does it (I mean, for each deployment for all PCs)?

**Comment by donho** [comment]
@gjibrane 
> There's no such option, but normal, Notepad++ MSI doesn't reboot, does it (I mean, for each deployment for all PCs)?

After googling, it seems SCCM triggers reboot. You can add `\norestart` as argument to prevent reboot.
However, I have added property "REBOOT" with value "supPress", and updated the MSI:
https://download.notepad-plus-plus.org/MISC/npp.Installer.x64.msi

**Comment by gjibrane** [comment]
@donho 
I made a second test with the first msi.
sccm triggers a reboot because of 1641 returncode.

I join the log

[npp.8.8.8.x64.log](https://github.com/user-attachments/files/23757885/npp.8.8.8.x64.log)

**Comment by donho** [comment]
Thank you for your log @gjibrane 
From the log, it seems NOAUUPDATER & REBOOT Suppress work.
Could you confirm that?

**Comment by gjibrane** [comment]
@donho 
Cannot confirm REBOOT in previous log because not used.

But made a new test with new MSI and REBOOT option.
It worked : 

[npp.8.8.8.x64.log](https://github.com/user-attachments/files/23758173/npp.8.8.8.x64.log)

I noticed that each 8.8.8 msi I downloaded have a different new GUID.
It makes an issue in detection method in sccm because all my instaals have different guid.

Maybe you should use a fixed custom guid for each version with version number present in it.

Other thing, along with nsis install I was using a config.model.xml :

[config.model.xml](https://github.com/user-attachments/files/23758225/config.model.xml)
Is it still a thing with MSI ? May be it was just to suppress update.

I was also using a nativeLang.xml to localize (French) by default.

And also was copying plugins because users can't download them.

Is there new way to proceed with MSI ?

**Comment by donho** [comment]
@gjibrane 
Thank you for the confirmation.

> I noticed that each 8.8.8 msi I downloaded have a different new GUID.
It makes an issue in detection method in sccm because all my instaals have different guid.

Isn't it rather a feature? As I understand, the GUID of MSI should be different so system knows it's a new update.

> Other thing, along with nsis install I was using a config.model.xml 

Could you tell me what's for this config.model.xml please?

> I was also using a nativeLang.xml to localize (French) by default.

"A piece of finger is given, then people ask the whole arm." :P 
I will see what I can do about it.

**Comment by gjibrane** [comment]
@donho 
> Isn't it rather a feature? As I understand, the GUID of MSI should be different so system knows it's a new update.

My problem comes from the fact I deployed for testing different 8.8.8 versions with different guid. It will probably be ok with normal releases.

> Could you tell me what's for this config.model.xml please?

I can't remember if there was other needs than disabling updater.
I will stop using it with msi. Moreover updated versions still keep their custom config.model.xml.
It could only be a problem with new install? Users will have to do their own settings. Not a real problem.

A parameter for setting LANGUAGE would be fine.

I'll manage plugins on my own as before, however in my humble opinion, "compare files" should be native since day one. :P

**Comment by alexhass** [comment]
> [@alexhass](https://github.com/alexhass) Thank you for your code. However, I managed to integrate it into current ws, compiled with WiX v6, and no compiling error:
> 
> 	<ComponentGroup Id="ContextMenuComponentGroup" Directory="ContextMenuFolder">
> 	  <Component Id="cmp_NppShellDll" Guid="{D6083500-DFD2-445D-B4FF-49EDF80A6A84}" Bitness="always64">
> 		<File Source="..\..\bin64\NppShell.x64.dll" Name="NppShell.dll" KeyPath="yes" >
>           <Class Id="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Context="InprocServer32" Description="Notepad++ Context Menu" ThreadingModel="apartment" />
>         </File>
> 		<RegistryValue Root="HKCR" Key="*\shellex\ContextMenuHandlers\Notepad++64" Value="{B298D29A-A6ED-11DE-BA8C-A68E55D89593}" Type="string" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Title" Value="Edit with &amp;Notepad++" Type="string" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Path" Value="[APPLICATIONFOLDER]notepad++.exe" Type="string" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Custom" Value="" Type="string" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="ShowIcon" Value="1" Type="integer" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Dynamic" Value="1" Type="integer" Action="write" />
>         <RegistryValue Root="HKCR" Key="CLSID\{B298D29A-A6ED-11DE-BA8C-A68E55D89593}\Settings" Name="Maxtext" Value="25" Type="integer" Action="write" />
> 	  </Component>
> 	  
> 	  <Component Id="cmp_NppShellMsix" Guid="{EB6E5606-60D1-4C69-8C9B-C0B12A19C6CE}">
> 		<File Source="..\..\bin64\NppShell.msix" KeyPath="yes" />
> 	  </Component>
> 	</ComponentGroup>
> However, under my Windows11, after the installation, there's no Notepad++ context menu entry at all. If you see any error in the above code cause the absent of context menu entry, feel free to let me know.

Have you added the feature to the wix package so it gets installed? I missed this in the example...:-(

**Comment by alexhass** [comment]
> [@donho](https://github.com/donho)
> 
> > Isn't it rather a feature? As I understand, the GUID of MSI should be different so system knows it's a new update.
> > My problem comes from the fact I deployed for testing different 8.8.8 versions with different guid. It will probably be ok with normal releases.
> 
> > Could you tell me what's for this config.model.xml please?
> > I can't remember if there was other needs than disabling updater.
> > I will stop using it with msi. Moreover updated versions still keep their custom config.model.xml.
> > It could only be a problem with new install? Users will have to do their own settings. Not a real problem.
> 
> A parameter for setting LANGUAGE would be fine.
> 
> I'll manage plugins on my own as before, however in my humble opinion, "compare files" should be native since day one. :P

What should LANGUAGE do? You need an option for every language... see my installer. Than use an MST for customizing and enable or disable the language you£d like to install... there are a LOT of languages.

**Comment by chris18890** [comment]
@donho Just wanted to say thank you for making the official MSI installer, it's very much appreciated! 

And to @alexhass for the unofficial MSI installer over the years, thank you as well for your efforts & helping prove there is a demand for this, as well as undoubtedly helping loads of IT Pros out there!

**Comment by m-p-3** [comment]
> Isn't it rather a feature? As I understand, the GUID of MSI should be different so system knows it's a new update.

It is, all the different installers should use a unique ProductCode, as a Detection Method it's simple and effective. That way, if there are multiple revision of the installer with the same version number, you can still determine which one was installed.

The only static part between the installers is the UpgradeCode.

**Comment by bluikko** [comment]
> > Isn't it rather a feature? As I understand, the GUID of MSI should be different so system knows it's a new update.
> 
> It is, all the different installers should use a unique ProductCode, as a Detection Method it's simple and effective. That way, if there are multiple revision of the installer with the same version number, you can still determine which one was installed.
> 
> The only static part between the installers is the UpgradeCode.

From operator point of view this is important to get right. The amount of MSIs that do not get it right is aggravating and then every time one adds a new version you need to manually list the old version(s) that are being updated and mistakes will happen.

**Comment by donho** [comment]
@alexhass 
> Have you added the feature to the wix package so it gets installed? I missed this in the example...:-(

Yes, I did.
Here are the source code files:
https://github.com/notepad-plus-plus/notepad-plus-plus/tree/master/PowerEditor/installer/msi

BTW, after the installation, I have verified the register with regedit.exe. All the keys you wanna write in the provided script is written.

**Comment by alexhass** [comment]
@donho It is really annoying to me that you never answer to my questions if you gave false arguments and I told you so. You never proved your arguments in last years and now about msi setups. You just say something random and cannot show what the problem is. And than just silence. Any chance you answer the questions?

Without a code review I cannot say why the context menu it does not work. It works for years in my code.

**Comment by alexhass** [comment]
@donho Any chance you answer the questions?

**Comment by donho** [comment]
@alexhass 
I think I did reply your questions.
If not, could you re-ask here please?

**Comment by donho** [comment]
@gjibrane 
> A parameter for setting LANGUAGE would be fine.

Here you go - use **NPP_LANG=french** to install the French localization:
PR: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/17946
For testing (Signed MSI): https://github.com/donho/notepad-plus-plus/releases/download/TEST2/npp.Installer.x64.msi

**Comment by gjibrane** [comment]
> [@gjibrane](https://github.com/gjibrane)
> 
> > A parameter for setting LANGUAGE would be fine.
> 
> Here you go - use **NPP_LANG=french** to install the French localization: PR: [#17946](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/17946) For testing (Signed MSI): https://github.com/donho/notepad-plus-plus/releases/download/TEST2/npp.Installer.x64.msi

Hello,
The download link is not working.

**Comment by alexhass** [comment]
There is HKCU inside. We are in an per machine installer. This is wrong. 

Additinally never run commands inside an MSI file. This will cause hang conditions and unstable setups. Do it in an MSI way, please.

**Comment by donho** [comment]
> There is HKCU inside. We are in an per machine installer. This is wrong.
> 
> Additinally never run commands inside an MSI file. This will cause hang conditions and unstable setups. Do it in an MSI way, please.

Fixed by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/09e50bad56c493a6dfd2d034d288b5271a656d96

**Comment by menglisch** [comment]
@donho  Just for reference: 09e50bad56c493a6dfd2d034d288b5271a656d96 caused issue #18067
